File: site/tools/csp.mjs
/**
 * Zet in elke gebouwde pagina een Content-Security-Policy als <meta>-tag.
 *
 * GitHub Pages kan geen eigen HTTP-headers meesturen, dus de meta-variant is
 * de enige manier om hier een CSP te krijgen. Die kent één beperking:
 * frame-ancestors werkt alleen als echte header en wordt in een meta-tag
 * genegeerd. Bescherming tegen clickjacking zit er dus niet in; dat lukt pas
 * met een proxy als Cloudflare ervoor.
 *
 * De inline scripts en stijlen krijgen een hash in plaats van 'unsafe-inline'.
 * Daarmee blijft een script dat een aanvaller in de pagina zou krijgen —
 * bijvoorbeeld via een projectnaam uit de beheeromgeving — alsnog geweigerd.
 * De hashes worden per pagina opnieuw berekend, zodat een wijziging aan een
 * script niet stilletjes de pagina breekt.
 *
 * Gebruik: draait automatisch na `npm run build`.
 */
import { createHash } from "node:crypto";
import { readFileSync, writeFileSync, globSync } from "node:fs";

const DIST = new URL("../dist/", import.meta.url).pathname;

/** sha256-hash in de vorm die een CSP verwacht. */
const hash = (inhoud) => `'sha256-${createHash("sha256").update(inhoud, "utf8").digest("base64")}'`;

/** Alle <script>- en <style>-inhoud uit een pagina, inclusief de JSON-LD-blokken. */
function inlineHashes(html, tag) {
  const uit = new Set();
  for (const m of html.matchAll(new RegExp(`<${tag}(?![^>]*\\ssrc=)[^>]*>([\\s\\S]*?)</${tag}>`, "g"))) {
    if (m[1] !== "") uit.add(hash(m[1]));
  }
  return [...uit];
}

function beleid(html) {
  const scripts = inlineHashes(html, "script");
  const stijlen = inlineHashes(html, "style");
  return [
    "default-src 'self'",
    "base-uri 'self'",
    "object-src 'none'",
    `script-src 'self' ${scripts.join(" ")}`.trim(),
    `style-src 'self' ${stijlen.join(" ")}`.trim(),
    "img-src 'self' data:",
    "font-src 'self'",
    // Het contact- en locatieformulier versturen naar Web3Forms.
    "connect-src 'self' https://api.web3forms.com",
    "form-action 'self' https://api.web3forms.com",
    // De video op de nieuwspagina komt van YouTube.
    "frame-src https://www.youtube.com https://www.youtube-nocookie.com",
    "upgrade-insecure-requests",
  ].join("; ");
}

let aangepast = 0;
let overgeslagen = 0;
for (const rel of globSync("**/*.html", { cwd: DIST })) {
  const pad = DIST + rel;
  let html = readFileSync(pad, "utf8");
  if (html.includes('http-equiv="Content-Security-Policy"')) continue;
  // De doorstuurpagina's van Astro zijn losse <meta refresh>-bestanden zonder
  // head en zonder script; daar valt niets te beveiligen.
  if (html.includes('http-equiv="refresh"')) { overgeslagen++; continue; }
  const tag = `<meta http-equiv="Content-Security-Policy" content="${beleid(html)}">`;
  // Zo vroeg mogelijk in de head, nog voor het eerste bestand wordt geladen.
  const na = html.indexOf("<head>");
  if (na === -1) {
    console.warn(`geen <head> gevonden, overgeslagen: ${rel}`);
    continue;
  }
  html = html.slice(0, na + 6) + "\n" + tag + html.slice(na + 6);
  writeFileSync(pad, html);
  aangepast++;
}
console.log(`CSP toegevoegd aan ${aangepast} pagina's (${overgeslagen} doorstuurpagina's overgeslagen).`);

File: tools/csp.py
#!/usr/bin/env python3
"""
Zet in elke pagina onder docs/ een Content-Security-Policy als <meta>-tag.

GitHub Pages stuurt geen eigen HTTP-headers mee, dus de meta-variant is hier
de enige mogelijkheid. Eén beperking hoort daarbij: frame-ancestors werkt
alleen als echte header en wordt in een meta-tag genegeerd. Bescherming tegen
clickjacking zit er dus niet in; dat lukt pas met een proxy als Cloudflare.

De inline scripts krijgen een hash in plaats van 'unsafe-inline'. Een script
dat een aanvaller in de pagina zou weten te krijgen, wordt daarmee alsnog
geweigerd. Voor stijl is dat niet haalbaar: de pagina's gebruiken honderden
style-attributen, en daar bestaat geen werkbare hash-variant voor.

Draai dit na elke wijziging aan de pagina's onder docs/:  python3 tools/csp.py
"""
import base64
import hashlib
import pathlib
import re

ROOT = pathlib.Path(__file__).resolve().parent.parent
DOCS = ROOT / "docs"

MERK = '<meta http-equiv="Content-Security-Policy"'

# Inline <script>-blokken zonder src. JSON-LD hoort er ook bij: een CSP kijkt
# naar het element, niet naar wat erin staat.
SCRIPTS = re.compile(r'<script(?![^>]*\ssrc=)[^>]*>(.*?)</script>', re.S)
HANDLERS = re.compile(r'\son(?:click|error|load|change|submit)="([^"]*)"')


def hash_van(inhoud: str) -> str:
    digest = hashlib.sha256(inhoud.encode("utf-8")).digest()
    return "'sha256-" + base64.b64encode(digest).decode("ascii") + "'"


def beleid(html: str, pad: str) -> str:
    scripts = dict.fromkeys(hash_van(m.group(1)) for m in SCRIPTS.finditer(html) if m.group(1))
    handlers = dict.fromkeys(hash_van(m.group(1)) for m in HANDLERS.finditer(html))

    script_src = ["'self'", *scripts]
    # Een onclick of onerror in de tag zelf draait niet op een gewone hash; dat
    # vraagt om 'unsafe-hashes' plus de hash van precies die ene regel code.
    if handlers:
        script_src.append("'unsafe-hashes'")
        script_src.extend(handlers)

    connect_src = ["'self'", "https://api.web3forms.com"]
    frame_src = ["https://www.youtube.com", "https://www.youtube-nocookie.com"]
    img_src = ["'self'", "data:", "https://img.youtube.com"]

    # De beheeromgeving haalt de Supabase-bibliotheek van een CDN en praat met
    # het Supabase-project zelf. Alleen die pagina krijgt die uitzondering.
    if pad.startswith("admin/"):
        script_src.append("https://cdn.jsdelivr.net")
        connect_src.extend(["https://*.supabase.co", "https://cdn.jsdelivr.net"])
        img_src.append("https://*.supabase.co")

    return "; ".join([
        "default-src 'self'",
        "base-uri 'self'",
        "object-src 'none'",
        "script-src " + " ".join(script_src),
        "style-src 'self' 'unsafe-inline'",
        "img-src " + " ".join(img_src),
        "font-src 'self'",
        "connect-src " + " ".join(connect_src),
        "form-action 'self' https://api.web3forms.com",
        "frame-src " + " ".join(frame_src),
        "upgrade-insecure-requests",
    ])


def main() -> None:
    aangepast = overgeslagen = vervangen = 0
    for bestand in sorted(DOCS.rglob("*.html")):
        html = bestand.read_text(encoding="utf-8")
        pad = str(bestand.relative_to(DOCS))

        if 'http-equiv="refresh"' in html:
            overgeslagen += 1
            continue

        # Een eerdere CSP eerst weghalen, zodat opnieuw draaien de hashes
        # bijwerkt in plaats van een tweede regel toe te voegen.
        nieuw = re.sub(r'\n?[ \t]*' + re.escape(MERK) + r'[^>]*>', "", html)
        if nieuw != html:
            vervangen += 1
        html = nieuw

        tag = f'  {MERK} content="{beleid(html, pad)}">'
        plek = html.find("<head>")
        if plek == -1:
            print(f"geen <head>, overgeslagen: {pad}")
            overgeslagen += 1
            continue

        html = html[: plek + 6] + "\n" + tag + html[plek + 6 :]
        bestand.write_text(html, encoding="utf-8")
        aangepast += 1

    print(f"CSP gezet in {aangepast} pagina's "
          f"({vervangen} bijgewerkt, {overgeslagen} overgeslagen).")


if __name__ == "__main__":
    main()
